### PR TITLE
Camel 14935 kafka exception

### PR DESCRIPTION
**Pull request by DariusX**
In KafkaConsumer, clear lastProcessedOffset if Exception on partition Revoked

### CODE REVIEW

File: components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaConsumer.java [hunk 1]
```diff
@@ -460,7 +460,14 @@ public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
...
                 }
                 LOG.debug("Saving offset repository state {} from offsetKey {} with offset: {}", threadId, offsetKey, offset);
-                commitOffset(offsetRepository, partition, offset, true);
+                try {
+                    commitOffset(offsetRepository, partition, offset, true);
+                } catch (java.lang.Exception e) {
+                    LOG.error("Error saving offset repository state {} from offsetKey {} with offset: {}", threadId, offsetKey, offset);
+                    throw e;
+                } finally {
+                    lastProcessedOffset.clear();
```
Comment (davsclaus) [inline components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaConsumer.java:469]: This finally block code is always exectued also for success. Is this really intended?

Comment (DariusX) [reply components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaConsumer.java:469]: Thank you for catching that. I'm going to decline the request and create a new one.

Comment (ceejay66) [reply components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaConsumer.java:469]: Yes removal is required on success.  This fixes an issue where an exception on commit prevents the map entrry being removed.

Comment (DariusX) [reply components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaConsumer.java:469]: Thank you. 
@davsclaus Yes, doing it in the finally is correct.

**Comment by DariusX** [comment]
@davsclaus  The code is fine, with the clearing in the finally block.
Can this move forward, please. Thx.

**Comment by davsclaus** [comment]
No its not correct, you clear the map even if there is NO exception.

File: components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaConsumer.java [hunk 2]
```diff
@@ -460,7 +460,14 @@ public void onPartitionsRevoked(Collection<TopicPartition> partitions) {
                     offset = -1L;
                 }
                 LOG.debug("Saving offset repository state {} from offsetKey {} with offset: {}", threadId, offsetKey, offset);
-                commitOffset(offsetRepository, partition, offset, true);
+                try {
+                    commitOffset(offsetRepository, partition, offset, true);
+                } catch (java.lang.Exception e) {
+                    LOG.error("Error saving offset repository state {} from offsetKey {} with offset: {}", threadId, offsetKey, offset);
```
Comment (davsclaus) [inline components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaConsumer.java:466]: What if you move the clear up here, and remove it from finally. Then its only cleared when there is an exception. And on success then the code below will only remove the current offset key, which seems correct to me.

Comment (ceejay66) [reply components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaConsumer.java:466]: We don't want to clear the map as it is used by multiple consumers and we will lose their state of last processed offset that may not be committed yet.  We just want to remove the entry from the map as it did originally but ensure we do this in the exception case as well as the non exception case. Moving the remove to the finally block gives us that behavior

**Comment by davsclaus** [comment]
@DariusX is this the same problem in camel-2.25.x branch? If so you are welcome to create a PR against this branch

**Comment by DariusX** [comment]
Yes, same in 2.25.x
Created new PR for that: https://github.com/apache/camel/pull/3884

**Pull request merged**